File: Frontend/src/api/userService.ts
import axiosInstance from "./axiosInstance";
import { buildUrl, config } from "@/lib/config";

export interface User {
  id: string | number;
  name: string;
  email: string;
}

export interface CreateUserPayload {
  name: string;
  email: string;
}

export interface InterviewSummary {
  interview_id: string;
  role: string;
  experience: string;
  difficulty: string;
  skills: string[];
  status: string;
  score: number | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface ResumeSummary {
  id: number;
  file_name: string;
  role: string;
  experience: string;
  skills: string[];
  source: string;
  created_at: string | null;
}

export interface UserOverview {
  user: {
    public_id: string;
    email: string;
    full_name: string | null;
    phone: string | null;
  };
  stats: {
    total_interviews: number;
    completed: number;
    /** Set up but never submitted — not necessarily still being taken. */
    unfinished: number;
    average_score: number | null;
    best_score: number | null;
  };
  /** Newest row of any status — often an abandoned setup. */
  latest_interview: InterviewSummary | null;
  /** Newest scored interview — what a user means by "my last interview". */
  latest_completed: InterviewSummary | null;
  /** Up to 5 newest scored interviews; compare with stats.completed for truncation. */
  recent_completed: InterviewSummary[];
  resumes: ResumeSummary[];
}

export interface ResumeSection {
  name: string;
  text: string;
}

export interface ResumeDetail {
  id: number;
  file_name: string;
  mime_type: string | null;
  role: string;
  experience: string;
  difficulty: string;
  skills: string[];
  source: string;
  created_at: string | null;
  chunk_count: number;
  sections: ResumeSection[];
  /** The original upload isn't retained, so there is nothing to download. */
  original_file_available: boolean;
}

export interface ResumeDeleteResult {
  deleted_id: number;
  file_name: string;
  was_vector_source: boolean;
  /** False when the vector store couldn't be reached; embeddings may linger. */
  vector_cleanup_ok: boolean;
  vectors_removed: number;
  job_setup_detached: boolean;
  message: string;
}

export const getResumeDetail = async (resumeId: number): Promise<ResumeDetail> => {
  const { data } = await axiosInstance.get<ResumeDetail>(
    buildUrl(config.ENDPOINTS.PROFILE.RESUME, { id: String(resumeId) })
  );
  return data;
};

export const deleteResume = async (resumeId: number): Promise<ResumeDeleteResult> => {
  const { data } = await axiosInstance.delete<ResumeDeleteResult>(
    buildUrl(config.ENDPOINTS.PROFILE.RESUME, { id: String(resumeId) })
  );
  return data;
};

export const getUserOverview = async (): Promise<UserOverview> => {
  const { data } = await axiosInstance.get<UserOverview>(
    config.ENDPOINTS.PROFILE.OVERVIEW
  );
  return data;
};

export const getUsers = async (): Promise<User[]> => {
  const { data } = await axiosInstance.get<User[]>(config.ENDPOINTS.USERS.LIST);
  return data;
};

export const createUser = async (payload: CreateUserPayload): Promise<User> => {
  const { data } = await axiosInstance.post<User>(config.ENDPOINTS.USERS.CREATE, payload);
  return data;
};

File: TalentPulseAI-fastAPI/app/services/user_service.py
"""
User-facing account data. Everything here is scoped to the authenticated user:
callers pass the User resolved from the JWT, never a client-supplied id.
"""
from typing import Dict

from sqlalchemy.orm import Session

from app.models.user import User
from app.services import interview_service, job_search_service


def get_overview(db: Session, user: User) -> Dict:
    """Profile-page payload: identity + interview history/status + indexed resumes."""
    interviews = interview_service.list_interviews(db, user.id)
    completed = [i for i in interviews if i["status"] == "submitted"]
    scores = [i["score"] for i in completed if isinstance(i["score"], (int, float))]
    # A row is created the moment the setup wizard finishes, so abandoned setups are
    # common and the newest row is often one of them. Surface the newest *scored*
    # interview separately — that is the one a user means by "my last interview".
    return {
        "user": {
            "public_id": user.public_id,
            "email": user.email,
            "full_name": user.full_name,
            "phone": user.phone,
        },
        "stats": {
            "total_interviews": len(interviews),
            "completed": len(completed),
            # Set up but never submitted — not necessarily still being taken.
            "unfinished": len(interviews) - len(completed),
            "average_score": round(sum(scores) / len(scores), 1) if scores else None,
            "best_score": max(scores) if scores else None,
        },
        "latest_interview": interviews[0] if interviews else None,
        "latest_completed": completed[0] if completed else None,
        # Scored interviews only — these are the ones with a report to open.
        # stats.completed is the true total, so the client can disclose truncation.
        "recent_completed": completed[:5],
        "resumes": job_search_service.list_resumes(db, user.id),
    }
